File: src-tauri/src/vault/images.rs
use crate::vault::resolve_vault_path;
use std::fs;

const IMAGES_DIR: &str = ".images";

static MIME_TO_EXT: &[(&str, &str)] = &[
    ("image/png", ".png"),
    ("image/jpeg", ".jpg"),
    ("image/gif", ".gif"),
    ("image/webp", ".webp"),
];

fn mime_to_ext(mime: &str) -> &'static str {
    MIME_TO_EXT
        .iter()
        .find(|(m, _)| *m == mime)
        .map(|(_, e)| *e)
        .unwrap_or(".png")
}

fn generate_image_filename(mime_type: &str) -> String {
    let ext = mime_to_ext(mime_type);
    let id = uuid::Uuid::new_v4().to_string().replace('-', "");
    format!("{}{}", &id[..12], ext)
}

/// Saves image bytes to `.images/` inside the notebook directory.
/// Returns the relative path for use in Markdown: `.images/abc123.png`
pub fn save_image(
    vault_path: &str,
    notebook: &str,
    image_data: &[u8],
    mime_type: &str,
) -> Result<String, String> {
    let resolved = resolve_vault_path(vault_path);
    let images_dir = resolved.join(notebook).join(IMAGES_DIR);
    fs::create_dir_all(&images_dir).map_err(|e| e.to_string())?;

    let filename = generate_image_filename(mime_type);
    let file_path = images_dir.join(&filename);
    fs::write(&file_path, image_data).map_err(|e| e.to_string())?;

    Ok(format!("{}/{}", IMAGES_DIR, filename))
}

/// Extracts image filenames referenced in Markdown content.
/// Matches `![alt](.images/filename)` patterns.
pub fn extract_image_references(markdown_content: &str) -> Vec<String> {
    use regex::Regex;
    let re = Regex::new(r"!\[[^\]]*\]\(\.images/([^)]+)\)").expect("valid regex");
    re.captures_iter(markdown_content)
        .filter_map(|cap| cap.get(1).map(|m| m.as_str().to_string()))
        .collect()
}

/// Moves image files from one notebook's `.images/` to another's.
pub fn move_images(
    vault_path: &str,
    source_notebook: &str,
    target_notebook: &str,
    filenames: &[String],
) -> Vec<String> {
    if filenames.is_empty() {
        return vec![];
    }

    let resolved = resolve_vault_path(vault_path);
    let src_dir = resolved.join(source_notebook).join(IMAGES_DIR);
    let dest_dir = resolved.join(target_notebook).join(IMAGES_DIR);

    let _ = fs::create_dir_all(&dest_dir);

    let mut moved = Vec::new();

    for filename in filenames {
        let src = src_dir.join(filename);
        let dest = dest_dir.join(filename);

        if src.exists() {
            if dest.exists() {
                moved.push(filename.clone());
                continue;
            }
            if fs::rename(&src, &dest).is_ok() {
                moved.push(filename.clone());
            }
        }
    }

    moved
}

File: electron/vault/images.ts
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import { resolveVaultPath } from './index';

const IMAGES_DIR = '.images';

const MIME_TO_EXT: Record<string, string> = {
  'image/png': '.png',
  'image/jpeg': '.jpg',
  'image/gif': '.gif',
  'image/webp': '.webp',
};

/**
 * Regex to extract image filenames from Markdown image references.
 * Matches patterns like: ![alt](.images/abc123.png)
 */
const IMAGE_REF_REGEX = /!\[[^\]]*\]\(\.images\/([^)]+)\)/g;

/**
 * Generates a unique filename with the correct extension based on MIME type.
 */
export function generateImageFilename(mimeType: string): string {
  const ext = MIME_TO_EXT[mimeType] ?? '.png';
  const id = crypto.randomUUID().replace(/-/g, '').slice(0, 12);
  return `${id}${ext}`;
}

/**
 * Saves image data to the .images/ subdirectory within the notebook.
 * Returns the relative path for use in Markdown references (e.g., ".images/abc123.png").
 */
export function saveImage(
  vaultPath: string,
  notebook: string,
  imageData: Buffer,
  mimeType: string
): string {
  const resolved = resolveVaultPath(vaultPath);
  const imagesDir = path.join(resolved, notebook, IMAGES_DIR);
  
  console.log(`[Electron images] Creating directory: ${imagesDir}`);
  fs.mkdirSync(imagesDir, { recursive: true });

  const filename = generateImageFilename(mimeType);
  const filePath = path.join(imagesDir, filename);
  
  console.log(`[Electron images] Writing image: ${filePath} (${imageData.length} bytes)`);
  fs.writeFileSync(filePath, imageData);
  console.log(`[Electron images] Image saved successfully: ${filename}`);

  // Return relative path for Markdown reference
  return `${IMAGES_DIR}/${filename}`;
}

/**
 * Resolves the full filesystem path for an image file.
 */
export function getImagePath(
  vaultPath: string,
  notebook: string,
  filename: string
): string {
  const resolved = resolveVaultPath(vaultPath);
  return path.join(resolved, notebook, IMAGES_DIR, filename);
}

/**
 * Extracts image filenames from Markdown content.
 * Finds all references matching ![alt](.images/filename) and returns the filenames.
 */
export function extractImageReferences(markdownContent: string): string[] {
  const filenames: string[] = [];
  let match: RegExpExecArray | null;
  // Reset regex state for each call
  const regex = new RegExp(IMAGE_REF_REGEX.source, IMAGE_REF_REGEX.flags);
  while ((match = regex.exec(markdownContent)) !== null) {
    filenames.push(match[1]);
  }
  return filenames;
}

/**
 * Moves image files from one notebook's .images/ directory to another.
 * Only moves the specified filenames. Creates the destination .images/ directory
 * if it doesn't exist. Skips files that don't exist in the source.
 *
 * Returns the list of filenames that were successfully moved.
 */
export function moveImages(
  vaultPath: string,
  sourceNotebook: string,
  targetNotebook: string,
  filenames: string[]
): string[] {
  if (filenames.length === 0) return [];

  const resolved = resolveVaultPath(vaultPath);
  const srcImagesDir = path.join(resolved, sourceNotebook, IMAGES_DIR);
  const destImagesDir = path.join(resolved, targetNotebook, IMAGES_DIR);

  // Create destination .images/ directory if needed
  if (filenames.length > 0) {
    fs.mkdirSync(destImagesDir, { recursive: true });
  }

  const moved: string[] = [];

  for (const filename of filenames) {
    const srcPath = path.join(srcImagesDir, filename);
    const destPath = path.join(destImagesDir, filename);

    try {
      if (fs.existsSync(srcPath)) {
        // If a file with the same name exists at destination, skip to avoid overwrite
        if (fs.existsSync(destPath)) {
          // File already exists at destination — no need to move
          moved.push(filename);
          continue;
        }
        fs.renameSync(srcPath, destPath);
        moved.push(filename);
      }
    } catch (err) {
      console.error(`Failed to move image ${filename}:`, err);
      // Continue with remaining files
    }
  }

  return moved;
}
